pkg/api/server/v1alpha2/records.go

```diff
@@ -142,125 +142,40 @@ func (s *Server) ListRecords(ctx context.Context, req *pb.ListRecordsRequest) (*
 	if req.GetParent() == "" {
 		return nil, status.Error(codes.InvalidArgument, "parent missing")
 	}
-	parent, _, err := result.ParseName(req.GetParent())
-	if err != nil {
-		return nil, err
-	}
-	if err := s.auth.Check(ctx, parent, auth.ResourceRecords, auth.PermissionList); err != nil {
-		return nil, err
-	}
 
-	userPageSize, err := pageSize(int(req.GetPageSize()))
+	// Authentication
+	parent, resultName, err := result.ParseName(req.GetParent())
 	if err != nil {
 		return nil, err
 	}
-
-	start, err := pageStart(req.GetPageToken(), req.GetFilter())
-	if err != nil {
+	if err := s.auth.Check(ctx, parent, auth.ResourceRecords, auth.PermissionList); err != nil {
 		return nil, err
 	}
 
-	sortOrder, err := orderBy(req.GetOrderBy())
+	recordsLister, err := lister.OfRecords(s.recordsEnv, parent, resultName, req)
 	if err != nil {
 		return nil, err
 	}
 
-	env, err := recordCEL()
-	if err != nil {
-		return nil, err
-	}
-	prg, err := celenv.ParseFilter(env, req.GetFilter())
-	if err != nil {
-		return nil, err
-	}
-	// Fetch n+1 items to get the next token.
-	out, err := s.getFilteredPaginatedSortedRecords(ctx, req.GetParent(), start, userPageSize+1, prg, sortOrder)
+	records, nextPageToken, err := recordsLister.List(ctx, s.db)
 	if err != nil {
 		return nil, err
 	}
 
-	// If we returned the full n+1 items, use the last element as the next page
-	// token.
-	var nextToken string
-	if len(out) > userPageSize {
-		next := out[len(out)-1]
-		var err error
-		nextToken, err = pagination.EncodeToken(next.GetId(), req.GetFilter())
+	// If we found no records, check if result exists so we can return NotFound
+	if len(records) == 0 {
+		_, err := getResultByParentName(s.db, parent, resultName)
 		if err != nil {
 			return nil, err
 		}
-		out = out[:len(out)-1]
 	}
 
 	return &pb.ListRecordsResponse{
-		Records:       out,
-		NextPageToken: nextToken,
+		Records:       records,
+		NextPageToken: nextPageToken,
 	}, nil
 }
 
-// getFilteredPaginatedSortedRecords returns the specified number of results that
-// match the given CEL program.
-func (s *Server) getFilteredPaginatedSortedRecords(ctx context.Context, parent, start string, pageSize int, prg cel.Program, sortOrder string) ([]*pb.Record, error) {
```

Comment:
:+1: to dropping what is likely the source of bugs my team is seeing with filtering.